odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1516 @@
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+from numbers import Number
+
+from odl.operator.operator import (
+    Operator, OperatorComp, OperatorLeftScalarMult, OperatorRightScalarMult,
+    OperatorRightVectorMult, OperatorSum)
+from odl.operator.default_ops import (ResidualOperator, IdentityOperator,
+                                      ConstantOperator, ScalingOperator)
+from odl.set.space import LinearSpaceVector
+from odl.solvers.advanced import (proximal_arg_scaling, proximal_translation,
+                                  proximal_quadratic_perturbation)
+
+
+# TODO: Add missing functionals here
+__all__ = ('Functional', 'ConvexConjugateArgScaling',
+           'FunctionalLeftScalarMult',
+           'ConvexConjugateFuncScaling', 'ConvexConjugateLinearPerturb',
+           'ConvexConjugateTranslation', 'TranslatedFunctional')
+
+
+class Functional(Operator):
+    """Implementation of a functional class.
+
+    Notes
+    -----
+    Note that the implementation of the functional class has assued that the
+    functionals are defined on a Hilbert space, i.e., that we consider
+    functionals
+
+    .. math::
+
+            f : X \\to F,
+
+    where :math:`X` is a Hilbert space and :math:`F` is a field of scalars
+    associated with :math:`X`. This has been done in order to simplify the
+    concept of *convex conjugate functional*. Since Hilbert spaces as selfdual
+    the convex conjugate functional is defined as
+
+    .. math::
+
+            f^* : X \\to F,
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
+
+    See, e.g., [Lue1969]_, [Roc1970]_ and [BC2011]_.
+    """
+
+    # TODO: Update doc above. What to write?
+
+    def __init__(self, domain, linear=False, smooth=False, concave=False,
+                 convex=False, grad_lipschitz=np.inf):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `Set`
+            The domain of this functional, i.e., the set of elements to
+            which this functional can be applied
+        linear : `bool`
+            If `True`, the functional is considered as linear. In this
+            case, ``domain`` and ``range`` have to be instances of
+            `LinearSpace`, or `Field`.
+        smooth : `bool`, optional
+            If `True`, assume that the functional is continuously
+            differentiable
+        concave : `bool`, optional
+            If `True`, assume that the functional is concave
+        convex : `bool`, optional
+            If `True`, assume that the functional is convex
+        grad_lipschitz : 'float', optional
+            The Lipschitz constant of the gradient.
+        """
+
+        Operator.__init__(self, domain=domain,
+                          range=domain.field, linear=linear)
+        self._is_smooth = bool(smooth)
+        self._is_convex = bool(convex)
+        self._is_concave = bool(concave)
```

Comment:
I have yet to see any code where these three bits of information (smooth, convex, concave) are really useful. With that I mean "more useful than gradient/convex conjugate being `NotImplementedError`". I have a bad feeling that these guys add complexity and corner cases we don't want to take into account all the time. In general, we shouldn't try to be too clever and "prevent" the user from doing something "stupid", which may be just fine but simply not recongnized by our "intelligent" system.
Also: users writing a functional will feel forced to decide these three properties. What if it's unknown, but the user still wants to use `gradient`? No reason to forbid it IMO.